enhancements/machine-config/mco-state-reporting.md

```diff
@@ -0,0 +1,453 @@
+but most of it simply lives in the MCO code rather than in data structures. We want to put these abstract "phases" of node operations as triggered by the MCO into a concrete data structure. 
+
+The MCO also has "day to day" operations on its MachineConfigController, MachineConfigDaemon, and MachineConfigOperator pods. Tracking these daily progression statuses is just as important as the upgrade status. However, this information does not belong in the upgrade-progression datatype. Currently, only the MCO team has the skillset to dive into the code and figure out what "phase" of the MCO causes a non-upgrade related error. The goal of this work is to make that information accessible and understandable to an everyday audience.
+
+### User Stories
+
+* As a <role>, I want to <take some action> so that I can <accomplish a
+goal>.
+
+### Goals
+
+* Make a MachineConfigState type that succinctly holds the upgrade data.
+* Have API load be as minimal as possible but augment the proper objects as needed.
+* Add State reporting to the operator machineconfiguration object, and add the MCO logic for updating these structs.
+* aggregate as much MCO related data into easily accessible places as possible.
+
+### Non-Goals
+
+* Make existing/new API objects difficult to understand.
+
+## Proposal
+
+Create a mechanism inside of the MCO to update the new MachineConfigState API type
+with data about a Node's progress during an upgrade and any errors that occur during those
+processes. Also use this mechanism to update the machineconfiguration object with everyday information.
+
+The data created by this MachineConfigState would look like the following:
+
+```console
+$ oc get mcs
+NAME             HEALTH    PROGRESS   NODE                          ACTION      ERROR
+upgrade-master   Healthy   Ready      ip-10-0-35-46.ec2.internal    NodeReady   
+upgrade-worker   Healthy   Ready      ip-10-0-21-151.ec2.internal   NodeReady   
+```
+
+Where the upgrade-worker and upgrade-master contain the nodes in each respective pool. The itemized entries contain the
+most recent reported state of a given node in the pool. Running oc describe will produce output like the following:
+
+```console
+$ oc describe mcs/uprade-worker
+  Most Recent State:
+    Name:    ip-10-0-21-151.ec2.internal
+    Phase:   NodeDraining
+    Reason:  NodeUpgradeInProgress
+    State:   UpgradeInProgress
+    Time:    2023-10-05T15:59:30Z
+  Progression History:
+    Name And State:    ip-10-0-21-151.ec2.internal:UpgradeInProgress
+    Phase:             NodeDraining
+    Reason:            NodeUpgradeInProgress
+    Name And State:  ip-10-0-21-151.ec2.internal:Ready
+    Phase:           Node is reporting a ready state
+    Reason:          NodeReady
+Events:              <none>
+```
+
+This output would be a little bit larger with some potential errors and of course roughly 3 node entries. The Progression History contains
```

Comment:
Some of the cluster may have 200 nodes in a pool. I am wondering how the state will look like for such pool when user has set maxUnavailable: 40 or something like that. Will the oc describe mcstate/worker be useful to track node update progress for such pools in a cluster?